src/main/java/com/google/cloud/tools/managedcloudsdk/install/SdkInstaller.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk.install;
+
+import com.google.cloud.tools.managedcloudsdk.MessageListener;
+import com.google.cloud.tools.managedcloudsdk.OsInfo;
+import com.google.cloud.tools.managedcloudsdk.Version;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+
+/** Install an SDK by downloading, extracting and if necessary installing. */
+public class SdkInstaller {
+
+  private final ListeningExecutorService executorService;
+  private final FileResourceProviderFactory fileResourceProviderFactory;
+  private final ExtractorFactory extractorFactory;
+  private final DownloaderFactory downloaderFactory;
+  private final InstallerFactory installerFactory;
+
+  /** Use {@link #newInstaller} to instantiate. */
+  SdkInstaller(
+      FileResourceProviderFactory fileResourceProviderFactory,
+      DownloaderFactory downloaderFactory,
+      ExtractorFactory extractorFactory,
+      InstallerFactory installerFactory,
+      ListeningExecutorService executorService) {
+    this.fileResourceProviderFactory = fileResourceProviderFactory;
+    this.executorService = executorService;
+    this.downloaderFactory = downloaderFactory;
+    this.extractorFactory = extractorFactory;
+    this.installerFactory = installerFactory;
+  }
+
+  /**
+   * Download a Cloud SDK asynchronously.
+   *
+   * @param messageListener a listener on installer output
+   * @return a path to the Cloud SDK home directory (wrapped in a listenable future)
+   */
+  public ListenableFuture<Path> downloadSdk(final MessageListener messageListener) {
```

Comment:
Should there be something that makes sure this is only called once, since it assumes a running executorService?

Or maybe throw an error if the submit is rejected?

Reply:
added in an illegalstateexception